File: swt/src/std.r/kwic.f
      INTEGER BUF(102)
      INTEGER GETLIN
      CALL READD0
10000 IF((GETLIN(BUF,-10).EQ.-1))GOTO 10001
        CALL PUTROT(BUF,-11)
      GOTO 10000
10001 CALL SWT
      END
      SUBROUTINE PUTROT(BUF,OUTFIL)
      INTEGER BUF(1)
      INTEGER OUTFIL
      INTEGER I,OUTFIL
      INTEGER DISCA0
      I=1
      GOTO 10004
10002 CONTINUE
10004 IF((BUF(I).EQ.138))GOTO 10003
      IF((BUF(I).EQ.0))GOTO 10003
        IF((BUF(I).EQ.160))GOTO 10005
        IF((BUF(I).EQ.137))GOTO 10005
          IF((DISCA0(BUF(I)).NE.0))GOTO 10006
            CALL ROTATE(BUF,I,OUTFIL)
10006     CONTINUE
10007       I=I+(1)
          IF((BUF(I).EQ.160))GOTO 10008
          IF((BUF(I).EQ.137))GOTO 10008
          IF((BUF(I).EQ.138))GOTO 10008
          IF((BUF(I).EQ.0))GOTO 10008
          GOTO 10007
10008     CONTINUE
10005   CONTINUE
10009   IF((BUF(I).EQ.160))GOTO 10011
        IF((BUF(I).EQ.137))GOTO 10011
        GOTO 10010
10011     I=I+(1)
        GOTO 10009
10010 GOTO 10002
10003 RETURN
      END
      SUBROUTINE ROTATE(BUF,N,OUTFIL)
      INTEGER BUF(1)
      INTEGER I,N,OUTFIL
      I=N
      GOTO 10014
10012 I=I+(1)
10014 IF((BUF(I).EQ.138))GOTO 10013
      IF((BUF(I).EQ.0))GOTO 10013
        CALL PUTCH(BUF(I),OUTFIL)
      GOTO 10012
10013 CALL PUTCH(:140,OUTFIL)
      I=1
      GOTO 10017
10015 I=I+(1)
10017 IF((I.GE.N))GOTO 10016
        CALL PUTCH(BUF(I),OUTFIL)
      GOTO 10015
10016 CALL PUTCH(138,OUTFIL)
      RETURN
      END
      INTEGER FUNCTION DISCA0(BUF)
      INTEGER BUF(1)
      INTEGER WORD(102)
      INTEGER I
      INTEGER SEARCH
      I=1
      GOTO 10020
10018 I=I+1
10020 IF((BUF(I).EQ.160))GOTO 10019
      IF((BUF(I).EQ.137))GOTO 10019
      IF((BUF(I).EQ.0))GOTO 10019
      IF((BUF(I).EQ.138))GOTO 10019
      IF((I.GE.102))GOTO 10019
        WORD(I)=BUF(I)
      GOTO 10018
10019 WORD(I)=0
      CALL MAPSTR(WORD,1)
      IF((SEARCH(WORD).NE.-1))GOTO 10021
        DISCA0=0
        GOTO 10022
10021   DISCA0=1
10022 RETURN
      END
      INTEGER FUNCTION SEARCH(WORD)
      INTEGER WORD(1)
      INTEGER I,J,K
      INTEGER COMPA0
      COMMON /KWDCOM/KWDPOS,KWDTE0,NUMKW0
      INTEGER KWDPOS(200)
      INTEGER KWDTE0(2048)
      INTEGER NUMKW0
      INTEGER AAAAA0
      I=1
      J=NUMKW0
      IF((I.GT.J))GOTO 10023
10024     K=(I+J)/2
          AAAAA0=COMPA0(KWDTE0(KWDPOS(K)),WORD)
          GOTO 10025
10026       I=K+1
          GOTO 10027
10028       SEARCH=KWDPOS(K)
            RETURN
10029       J=K-1
          GOTO 10027
10025     GOTO(10026,10028,10029),AAAAA0
10027   CONTINUE
        IF((I.LE.J))GOTO 10024
10023 SEARCH=-1
      RETURN
      END
      INTEGER FUNCTION COMPA0(STR1,STR2)
      INTEGER STR1(1),STR2(1)
      INTEGER I
      I=1
      GOTO 10032
10030 I=I+1
10032 IF((STR1(I).NE.STR2(I)))GOTO 10031
        IF(((STR1(I).NE.0).AND.(STR2(I).NE.0)))GOTO 10033
          GOTO 10031
10033 GOTO 10030
10031 IF((STR1(I).NE.STR2(I)))GOTO 10034
        COMPA0=2
        GOTO 10035
10034   IF((STR1(I).EQ.0))GOTO 10037
        IF(((STR1(I).LT.STR2(I)).AND.(STR2(I).NE.0)))GOTO 10037
        GOTO 10036
10037     COMPA0=1
          GOTO 10038
10036     IF((STR2(I).EQ.0))GOTO 10040
          IF(((STR2(I).LT.STR1(I)).AND.(STR1(I).NE.0)))GOTO 10040
          GOTO 10039
10040       COMPA0=3
            GOTO 10041
10039       CALL ERROR('in compare: can''t happen.')
10041   CONTINUE
10038 CONTINUE
10035 RETURN
      END
      SUBROUTINE READD0
      INTEGER I,L,FD
      INTEGER GETLI0,GETLIN
      INTEGER LINE(102)
      COMMON /KWDCOM/KWDPOS,KWDTE0,NUMKW0
      INTEGER KWDPOS(200)
      INTEGER KWDTE0(2048)
      INTEGER NUMKW0
      NUMKW0=0
      IF((GETLI0(FD).NE.-3))GOTO 10042
        RETURN
10042 I=1
      L=GETLIN(LINE,FD)
10043 IF((L.EQ.-1))GOTO 10044
        IF((LINE(L).NE.138))GOTO 10045
          LINE(L)=0
          L=L-1
10045   IF((2048-I.LE.L))GOTO 10046
        IF((NUMKW0.GE.200))GOTO 10046
          NUMKW0=NUMKW0+1
          KWDPOS(NUMKW0)=I
          CALL SCOPY(LINE,1,KWDTE0,I)
          I=I+L+1
          GOTO 10047
10046     CALL PRINT(-15,'*s: discard list too long*n.',LINE)
          GOTO 10044
10047   L=GETLIN(LINE,FD)
      GOTO 10043
10044 IF((FD.EQ.-12))GOTO 10048
        CALL CLOSE(FD)
10048 RETURN
      END
      INTEGER FUNCTION GETLI0(FD)
      INTEGER FD
      INTEGER ARG(128)
      INTEGER GETARG,OPEN
      IF((GETARG(1,ARG,128).EQ.-1))GOTO 10049
        IF((ARG(1).NE.173))GOTO 10050
        IF(((ARG(2).NE.228).AND.(ARG(2).NE.196)))GOTO 10050
          IF((GETARG(2,ARG,128).EQ.-1))GOTO 10051
            FD=OPEN(ARG,1)
            IF((FD.NE.-3))GOTO 10052
              CALL CANT(ARG)
10052       GOTO 10053
10051       FD=-12
10053     GOTO 10054
10050     CALL ERROR('Usage: kwic [ -d [ <discard_list> ] ].')
10054   GOTO 10055
10049   FD=-3
10055 GETLI0=FD
      RETURN
      END
C ---- Long Name Map ----
C compare                        compa0
C getlist                        getli0
C discard                        disca0
C numkwds                        numkw0
C readdiscards                   readd0
C kwdtext                        kwdte0

File: swt/MFD/SRC/STD.R/KWIC.R
# kwic --- make keyword in context index

   define(FOLD,:140)
   define(MAX_KWDS,200)
   define(MAX_KWD_LEN,2048)

   character buf (MAXLINE)
   integer getlin

   call read_discards
   while (getlin (buf, STDIN) ~= EOF)
      call putrot (buf, STDOUT)

   stop
   end

# putrot --- create lines with keyword at front
   subroutine putrot (buf, outfil)
   character buf (ARB)
   integer outfil

   integer i, outfil
   integer discard

   for (i = 1; buf (i) ~= NEWLINE && buf (i) ~= EOS; ) {
      if (buf (i) ~= ' 'c && buf (i) ~= TAB) {  # start of a token
         if (discard (buf (i)) == NO)     # not a throw-away
            call rotate (buf, i, outfil)
         repeat
            i += 1
            until (buf (i) == ' 'c || buf (i) == TAB ||
                        buf (i) == NEWLINE || buf (i) == EOS)
         }
      while (buf (i) == ' 'c || buf (i) == TAB)
         i += 1
      }

   return
   end

# rotate --- output rotated line
   subroutine rotate (buf, n, outfil)
   character buf (ARB)
   integer i, n, outfil

   for (i = n; buf (i) ~= NEWLINE && buf (i) ~= EOS; i += 1)
      call putch (buf (i), outfil)
   call putch (FOLD, outfil)
   for (i = 1; i < n; i += 1)
      call putch (buf (i), outfil)
   call putch (NEWLINE, outfil)

   return
   end

# discard --- see if first word in buf is in throw-away list
   integer function discard (buf)
   character buf (ARB)

   character word (MAXLINE)
   integer i
   integer search

   for (i = 1; buf (i) ~= ' 'c && buf (i) ~= TAB && buf (i) ~= EOS &&
                  buf (i) ~= NEWLINE && i < MAXLINE; i = i + 1)
      word (i) = buf (i)

   word (i) = EOS
   call mapstr (word, LOWER)

   if (search (word) == EOF)
      discard = NO   # word not in list, don't throw it away
   else
      discard = YES

   return
   end

# search --- search keyword list for word
   integer function search (word)
   character word (ARB)

   integer i, j, k
   integer compare

   common /kwdcom/ kwd_pos, kwd_text, num_kwds
   integer kwd_pos (MAX_KWDS)       # keyword positions
   character kwd_text (MAX_KWD_LEN) # text of keywords
   integer num_kwds                 # number of keywords

   i = 1
   j = num_kwds
   if (i <= j)
      repeat { # perform binary search of the throw-away list
         k = (i + j) / 2
         select (compare (kwd_text (kwd_pos (k)), word))
            when (1)    # less
               i = k + 1
            when (2) {  # equal
               search = kwd_pos (k)
               return
               }
            when (3)    # greater
               j = k - 1
         } until (i > j)

   search = EOF
   return

   end

# compare --- compare two strings and return 1 2 or 3 for < = or >
   integer function compare (str1, str2)
   character str1 (ARB), str2 (ARB)

   integer i

   for (i = 1; str1 (i) == str2 (i); i = i + 1)
      if (str1 (i) == EOS | str2 (i) == EOS)
         break

   if (str1 (i) == str2 (i))                       # equal strings
      compare = 2

   else if (str1 (i) == EOS
     || str1 (i) < str2 (i) & str2 (i) ~= EOS)    # str1 < str2
      compare = 1

   else if (str2 (i) == EOS
     || str2 (i) < str1 (i) & str1 (i) ~= EOS)    # str1 > str2
      compare = 3

   else
      call error ("in compare: can't happen.")

   return
   end

# read_discards --- load throw-away list from file
   subroutine read_discards

   integer i, l, fd
   integer get_list, getlin

   character line (MAXLINE)

   common /kwdcom/ kwd_pos, kwd_text, num_kwds
   integer kwd_pos (MAX_KWDS)       # keyword positions
   character kwd_text (MAX_KWD_LEN) # text of keywords
   integer num_kwds                 # number of keywords

   num_kwds = 0   # initially zero
   if (get_list (fd) == ERR)
      return

   i = 1
   l = getlin (line, fd)
   while (l ~= EOF) {
      if (line (l) == NEWLINE) {
         line (l) = EOS
         l = l - 1
         }
      if (MAX_KWD_LEN - i > l && num_kwds < MAX_KWDS) { # check for room
         num_kwds = num_kwds + 1
         kwd_pos (num_kwds) = i
         call scopy (line, 1, kwd_text, i)
         i = i + l + 1
         }
      else {
         call print (ERROUT, "*s: discard list too long*n.", line)
         break
         }
      l = getlin (line, fd)
      }

   if (fd ~= STDIN2)
      call close (fd)

   return
   end

# get_list --- get throw-away list file descriptor
   integer function get_list (fd)
   integer fd

   character arg (MAXARG)
   integer getarg, open

   if (getarg (1, arg, MAXARG) ~= EOF)
      if (arg (1) == '-'c && (arg (2) == 'd'c | arg (2) == 'D'c)) {
         if (getarg (2, arg, MAXARG) ~= EOF) {
            fd = open (arg, READ)
            if (fd == ERR)
               call cant (arg)
            }
         else
            fd = STDIN2
         }
      else
         call error ("Usage: kwic [ -d [ <discard_list> ] ].")
   else
      fd = ERR

   get_list = fd

   return
   end
